File: csharp/day14_part1_2016.cs

using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

class Program
{
    static void Main()
    {
        string salt = File.ReadAllText("input.txt").Trim();
        int keys = 0;
        int index = 0;
        while (keys < 64)
        {
            string hash = GetMD5Hash(salt + index);
            string triplet = FindTriplet(hash);
            if (!string.IsNullOrEmpty(triplet))
            {
                for (int i = 1; i <= 1000; i++)
                {
                    string nextHash = GetMD5Hash(salt + (index + i));
                    if (nextHash.Contains(new string(triplet[0], 5)))
                    {
                        keys++;
                        break;
                    }
                }
            }
            index++;
        }
        Console.WriteLine(index - 1);
    }

    static string GetMD5Hash(string input)
    {
        using (MD5 md5 = MD5.Create())
        {
            byte[] inputBytes = Encoding.ASCII.GetBytes(input);
            byte[] hashBytes = md5.ComputeHash(inputBytes);
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < hashBytes.Length; i++)
            {
                sb.Append(hashBytes[i].ToString("x2"));
            }
            return sb.ToString();
        }
    }

    static string FindTriplet(string hash)
    {
        for (int i = 0; i < hash.Length - 2; i++)
        {
            if (hash[i] == hash[i + 1] && hash[i] == hash[i + 2])
            {
                return hash[i].ToString();
            }
        }
        return null;
    }
}

File: fsharp/day14_part1_2016.fs
open System
open System.IO
open System.Security.Cryptography
open System.Text

let getMd5Hash (input:string) =
    use md5 = MD5.Create()
    let bytes = Encoding.ASCII.GetBytes(input)
    let hashBytes = md5.ComputeHash(bytes)
    let sb = StringBuilder()
    for b in hashBytes do sb.Append(b.ToString("x2")) |> ignore
    sb.ToString()

let findTriplet (hash:string) =
    let rec loop i =
        if i > hash.Length - 3 then None
        elif hash.[i] = hash.[i+1] && hash.[i] = hash.[i+2] then Some hash.[i]
        else loop (i+1)
    loop 0

[<EntryPoint>]
let main _ =
    let salt = File.ReadAllText("input.txt").Trim()
    let mutable keys = 0
    let mutable index = 0
    while keys < 64 do
        let hash = getMd5Hash (salt + string index)
        match findTriplet hash with
        | None -> ()
        | Some c ->
            let quintuple = string c |> fun s -> s.PadRight(5, c)
            let mutable found = false
            for i in 1 .. 1000 do
                if not found then
                    let nextHash = getMd5Hash (salt + string (index + i))
                    if nextHash.Contains(quintuple) then
                        found <- true
                        keys <- keys + 1
        index <- index + 1
    printfn "%d" (index - 1)
    0